### PR TITLE
!htp #17965 more consistent and 22-ified javadsl handleWith

### PR DESCRIPTION
**Pull request by ktoso**
Adjusts javadsl as we discussed in #17965

Please review @2m @jrudolph

### CODE REVIEW

File: akka-docs-dev/rst/java/code/docs/http/javadsl/HandlerExampleDocTest.java [hunk 1]
```diff
@@ -12,7 +12,7 @@
 import akka.http.javadsl.testkit.TestRoute;
 import org.junit.Test;
 
-public class HandlerExampleSpec extends JUnitRouteTest {
+public class HandlerExampleDocTest extends JUnitRouteTest {
```
Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/HandlerExampleDocTest.java]: changed a few of those to Test. In general "java tests" = Test, "scala test" = Spec.

File: akka-docs-dev/rst/java/http/routing-dsl/handlers.rst [hunk 2]
```diff
@@ -59,16 +59,32 @@ route structure, this time representing segments from the URI path.
...
 
+.. note::
+  The reason the ``handleWith##`` methods include the number of handled values is because otherwise (if overloading would
+  be used, for all 22 methods) error messages generated by ``javac`` end up being very long and not readable, i.e.
+  if one type of a handler does not match the given values, *all* possible candidates would be printed in the error message
+  (22 of them), instead of just the one arity-matching method, pointing out that the type does not match.
+
+   We opted for better error messages as we feel this is more helpful when developing applications,
+   instead of having one overloaded method which looks nice when everything works, but procudes hard to read error
+   messages if something doesnt match up.
```
Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/handlers.rst]: felt the need to explain it somehow... keep or drop?

File: akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  private[http] def handle(extractions: RequestVal[_]*)(f: RequestContext ⇒ RouteResult): Route
+
+  /**
+   * Handles the route using the given function.
+   * The function MUST NOT return `null`.
+   *
+   * If the `handler` is accessing request values these must be passed to this method in order for extraction to be performed.
+   */
+  @varargs def handleWith(handler: akka.japi.function.Function[RequestContext, RouteResult], extractions: RequestVal[_]*): Route =
+    handle(extractions: _*)(ctx => handler(ctx))
```
Comment (ktoso) [inline akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template]: that's the "not typesafe handler"

File: akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  /**
+   * Handles the route using the given function, completing the route once the returned [[scala.concurrent.Future]] completes.
+   * The function MUST NOT return `null`.
+   *
+   * For convenience, using Java 8 lambda expressions as the `handler` function is recommended.
+   * For Java 6 or 7 users the convenience [[akka.http.javadsl.server.Handler1]] class (which itself extends
+   * [[akka.japi.function.Function2]] should prove to be useful, as it matches naming-wise with the number of
+   * handled request values.
+   */
+  def handleWithAsync1[[#T1#]]([#v1: RequestVal[T1]#], handler: akka.japi.function.Function2[RequestContext, [#T1#], Future[RouteResult]]): Route =
```
Comment (ktoso) [inline akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template]: needs an additional method for Async. Can't use overloads because erasure.

File: akka-docs-dev/rst/java/http/routing-dsl/handlers.rst (hunk 2 above)
Comment (jrudolph) [inline akka-docs-dev/rst/java/http/routing-dsl/handlers.rst]: I think it's nice.

s/doesnt/doesn't/

File: akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template (hunk 4 above)
Comment (jrudolph) [inline akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template]: Actually, you don't really need the async variant because there's `RequestContext.completeWith` which turns an async result into a (seemingly) sync one.

File: akka-docs-dev/rst/java/http/routing-dsl/handlers.rst (hunk 2 above)
Comment (jrudolph) [inline akka-docs-dev/rst/java/http/routing-dsl/handlers.rst]: Is the indenting correct (hard to see in the diff)?

File: akka-docs-dev/rst/java/http/routing-dsl/handlers.rst [hunk 5]
```diff
@@ -59,16 +59,32 @@ route structure, this time representing segments from the URI path.
...
+
+
 Providing Handlers by Reflection
 --------------------------------
 
-Using Java before Java 8, writing out handlers as (anonymous) classes can be unwieldy. Therefore, ``handleWith``
+Using Java before Java 8, writing out handlers as (anonymous) classes can be unwieldy. Therefore, ``handleWithReflect``
```
Comment (jrudolph) [inline akka-docs-dev/rst/java/http/routing-dsl/handlers.rst]: `handleWithReflect` sounds abbreviated. How about `handleReflectively` or `handleWithReflection`?

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/SchemeDirectivesSpec.scala [hunk 6]
```diff
@@ -22,10 +20,12 @@ class SchemeDirectivesSpec extends RoutingSpec {
       Get("https://localhost/") ~> scheme("http") { completeOk } ~> check { rejections shouldEqual List(SchemeRejection("http")) }
     }
     "cancel SchemeRejection if other scheme passed" in {
-      val route =
+      val route: Route =
         scheme("https") { completeOk } ~
           scheme("http") { reject }
 
+      akka.http.scaladsl.Http().bindAndHandle(route, "")
```
Comment (jrudolph) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/SchemeDirectivesSpec.scala]: ?

File: project/AkkaBuild.scala [hunk 7]
```diff
@@ -1037,8 +1037,8 @@ object AkkaBuild extends Build {
     Protobuf.settings ++ Seq(
     // compile options
     scalacOptions in Compile ++= Seq("-encoding", "UTF-8", "-target:jvm-1.6", "-deprecation", "-feature", "-unchecked", "-Xlog-reflective-calls", "-Xlint"),
-    javacOptions in compile ++= Seq("-encoding", "UTF-8", "-source", "1.6", "-target", "1.6", "-Xlint:unchecked", "-Xlint:deprecation"),
-    javacOptions in doc ++= Seq("-encoding", "UTF-8", "-source", "1.6") ++
+    javacOptions in compile ++= Seq("-encoding", "UTF-8", "-source", "1.8", "-target", "1.8", "-Xlint:unchecked", "-Xlint:deprecation"),
+    javacOptions in doc ++= Seq("-encoding", "UTF-8", "-source", "1.8") ++
```
Comment (jrudolph) [inline project/AkkaBuild.scala]: Was that change intended?

File: akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/SchemeDirectivesSpec.scala (hunk 6 above)
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/SchemeDirectivesSpec.scala]: ouch, thanks for noticing

File: akka-docs-dev/rst/java/http/routing-dsl/handlers.rst (hunk 2 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/handlers.rst]: indent is ok, fixed typo; it's a keeper then

File: akka-docs-dev/rst/java/http/routing-dsl/handlers.rst (hunk 5 above)
Comment (2m) [inline akka-docs-dev/rst/java/http/routing-dsl/handlers.rst]: +1 for `handleReflectively`

File: akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala [hunk 8]
```diff
@@ -567,7 +570,7 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
     new HttpExt(system.settings.config getConfig "akka.http")(system)
 }
 
-import JavaConverters._
+import scala.collection.JavaConverters._
```
Comment (2m) [inline akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala]: Looks like all of the changes in this file are not necessary.

File: akka-http-tests-java8/src/main/java/akka/http/javadsl/server/examples/simple/SimpleServerApp8.java [hunk 9]
```diff
@@ -63,25 +63,25 @@ public RouteResult handle(RequestContext ctx, Integer xVal, Integer yVal) {
...
                 // matches paths like this: /multiply/{x}/{y}
                 path("multiply", xSegment, ySegment).route(
                     // bind handler by reflection
-                    handleWith(xSegment, ySegment, SimpleServerApp8::multiply)
+                    handleWith2(xSegment, ySegment, SimpleServerApp8::multiply)
                 ),
                 path("multiply-methodref", xSegment, ySegment).route(
-                    // bind handler by reflection
-                    handleWith(xSegment, ySegment, new Test(123)::constantPlusMultiply)
+                    // bind handler by referende to new instance of handler
```
Comment (2m) [inline akka-http-tests-java8/src/main/java/akka/http/javadsl/server/examples/simple/SimpleServerApp8.java]: s/referende/reference

File: akka-http/src/main/boilerplate/akka/http/javadsl/server/Handlers.scala.template [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * A route Handler that handles a request (that is encapsulated in a [[RequestContext]])
+ * and returns a [[scala.concurrent.Future]] of [[RouteResult]] with the response (or the rejection).
+ *
+ * Use the methods in [[RequestContext]] to create a [[RouteResult]].
+ * A handler MUST NOT return `null` as the result.
+ */
+trait AsyncHandler extends akka.japi.function.Function[RequestContext, RouteResult] {
+  override def apply(ctx: RequestContext): RouteResult
+}
```
Comment (2m) [inline akka-http/src/main/boilerplate/akka/http/javadsl/server/Handlers.scala.template]: Should the return type of this be `Future[RouteResult]`?

File: akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template (hunk 4 above)
Comment (ktoso) [inline akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template]: Awesome, deleting the Async versions.

File: akka-docs-dev/rst/java/http/routing-dsl/handlers.rst (hunk 5 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/handlers.rst]: I also like `handleReflectively`, changing

File: akka-http-tests-java8/src/main/java/akka/http/javadsl/server/examples/simple/SimpleServerApp8.java (hunk 9 above)
Comment (ktoso) [inline akka-http-tests-java8/src/main/java/akka/http/javadsl/server/examples/simple/SimpleServerApp8.java]: ok

File: project/AkkaBuild.scala (hunk 7 above)
Comment (2m) [inline project/AkkaBuild.scala]: Why this change?

Comment (ktoso) [inline project/AkkaBuild.scala]: thanks; no, this is not needed (now - it was when I was playing around a while ago)

File: akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template (hunk 4 above)
Comment (ktoso) [inline akka-http/src/main/boilerplate/akka/http/javadsl/server/directives/BasicDirectivesBase.scala.template]: (brought them back again after some more thuoght_

File: akka-docs-dev/rst/java/http/routing-dsl/handlers.rst.orig [hunk 11]
```diff
@@ -0,0 +1,109 @@
+.. _handlers-java:
```
Comment (2m) [inline akka-docs-dev/rst/java/http/routing-dsl/handlers.rst.orig]: This file is merge left over.

File: akka-docs-dev/rst/scala/stream-quickstart.rst.orig [hunk 12]
```diff
@@ -0,0 +1,210 @@
+.. _stream-quickstart-scala:
```
Comment (2m) [inline akka-docs-dev/rst/scala/stream-quickstart.rst.orig]: Merge leftover as well.

**Comment by 2m** [comment]
There are quite a few `*.orig` files in this PR which should be removed.

**Comment by ktoso** [comment]
Ouch, thanks :<

**Comment by ktoso** [comment]
Pushed final cleanup. No orig files, no AkkaBuild touched

**Comment by ktoso** [comment]
Failure was a method reference in java6 sources, missed it :(

**Comment by ktoso** [comment]
Error was unrelated, timing issue in FlowTest, I'll open an issue to harden it.

File: akka-http/src/main/boilerplate/akka/http/javadsl/server/Handlers.scala.template (hunk 10 above)
Comment (ktoso) [inline akka-http/src/main/boilerplate/akka/http/javadsl/server/Handlers.scala.template]: thanks, is fixed
